core/src/main/java/tc/oc/pgm/filters/MonostableFilter.java

```diff
@@ -0,0 +1,112 @@
+package tc.oc.pgm.filters;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.Map;
+import tc.oc.pgm.api.filter.Filter;
+import tc.oc.pgm.api.filter.ReactorFactory;
+import tc.oc.pgm.api.filter.query.MatchQuery;
+import tc.oc.pgm.api.filter.query.Query;
+import tc.oc.pgm.api.match.Match;
+import tc.oc.pgm.api.match.MatchScope;
+import tc.oc.pgm.api.match.Tickable;
+import tc.oc.pgm.api.time.Tick;
+import tc.oc.pgm.filters.dynamic.FilterMatchModule;
+import tc.oc.pgm.filters.dynamic.Filterable;
+import tc.oc.pgm.filters.dynamic.Filterables;
+
+public class MonostableFilter extends SingleFilterFunction implements ReactorFactory {
+
+  private final Duration duration;
+  final Class<? extends Filterable<?>> scope;
+
+  // Filterables that currently pass the inner filter, mapped to the instants that they expire.
+  // They are not actually removed until the inner filter goes false.
+  final Map<Filterable<?>, Instant> endTimes = new HashMap<>();
+
+  public static Filter afterMatchStart(Duration duration) {
+    return AllFilter.of(after(MatchPhaseFilter.RUNNING, duration), MatchPhaseFilter.RUNNING);
+  }
+
+  /**
+   * Will rise after {@code duration} has passed after {@code filter} starts to rise.
+   *
+   * @param filter the filter to listen for changes to
+   * @param duration the duration to delay this filters rise
+   */
+  public static Filter after(Filter filter, Duration duration) {
+    return new InverseFilter(new MonostableFilter(filter, duration));
+  }
+
+  public MonostableFilter(Filter filter, Duration duration) {
+    super(filter);
+    this.duration = duration;
+    this.scope = Filterables.scope(filter);
+  }
+
+  @Override
+  public QueryResponse query(Query query) {
+    boolean response = this.filter.response(query);
+    if (!(query instanceof MatchQuery)) return QueryResponse.ABSTAIN; // TypedFilter<MatchQuery>
+    final MatchQuery matchQuery = ((MatchQuery) query);
+    final Filterable<?> filterable =
+        Filterables.extractFilterable(matchQuery).getFilterableAncestor(this.scope);
+    if (filterable == null)
+      throw new IllegalArgumentException(
+          "The scope of this filter does not match the query it received");
+
+    final Reactor reactor =
+        matchQuery
+            .getMatch()
+            .needModule(FilterMatchModule.class)
+            .getReactorInstance(MonostableFilter.Reactor.class);
```

Comment:
I'm confused by this

Is there ONE reactor for all monostable filters, or one reactor for EACH monostable filter?

If there is one global reactor for all monostable filters, then this makes sense, but "createReactor" documentation is wrong, as it states that createReactor is called for all filters, not just for "one random monostable filter that happens to be the first".

I think what makes most sense is for each monostable to have its own reactor, which stores state for that filter in that match, and instead of `getReactorInstance(Class)` have a proper `getReactorInstance(ReactorFactory)` and pass `this` as parameter

Reply:
It seems like my quick draft didn't make sense after all.. :cry:

Ill make sure to actually scope them to the filter and not the reactor type..